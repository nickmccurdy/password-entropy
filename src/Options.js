import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'

export default class Options extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    onChange: PropTypes.func.isRequired,
    possiblePasswords: PropTypes.func.isRequired
  }

  componentDidMount = this.change
  componentDidUpdate = this.change

  change () {
    this.props.onChange(this.props.possiblePasswords())
  }

  render () {
    return <div>{this.props.children}</div>
  }
}
