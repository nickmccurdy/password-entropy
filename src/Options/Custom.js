import { faQuestionCircle } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import PropTypes from 'prop-types'
import { FormGroup, Input, Label } from 'reactstrap'

export default function Custom({ custom, onCustom }) {
  return (
    <FormGroup>
      <Label>
        <FontAwesomeIcon icon={faQuestionCircle} /> Possible Items
        <Input
          value={custom}
          onChange={onCustom}
          type="number"
          min="0"
          required
        />
      </Label>
    </FormGroup>
  )
}

Custom.propTypes = {
  custom: PropTypes.number.isRequired,
  onCustom: PropTypes.func.isRequired,
}
