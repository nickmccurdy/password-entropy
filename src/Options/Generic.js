import {
  Checkbox,
  FormControlLabel,
  FormGroup,
  Typography
} from '@material-ui/core'
import PropTypes from 'prop-types'
import React, { Fragment } from 'react'

export const examples = {
  'Lowercase Letters': 'a-z',
  'Uppercase Letters': 'A-Z',
  Numbers: '0-9',
  Symbols: '!@#$%^&*'
}

export default function Generic({ onGeneric, toggles }) {
  return (
    <FormGroup>
      {Object.entries(examples).map(([name, example]) => (
        <FormControlLabel
          key={name}
          control={
            <Checkbox
              name={name}
              checked={toggles[name]}
              onChange={onGeneric}
            />
          }
          label={
            <Fragment>
              {name}

              <Typography variant="caption" gutterBottom>
                {example}
              </Typography>
            </Fragment>
          }
        />
      ))}
    </FormGroup>
  )
}

Generic.propTypes = {
  onGeneric: PropTypes.func.isRequired,
  toggles: PropTypes.shape({
    'Lowercase Letters': PropTypes.bool.isRequired,
    'Uppercase Letters': PropTypes.bool.isRequired,
    Numbers: PropTypes.bool.isRequired,
    Symbols: PropTypes.bool.isRequired
  }).isRequired
}
