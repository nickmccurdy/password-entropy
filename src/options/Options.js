import startCase from 'lodash/startCase'
import entries from 'object.entries'
import PropTypes from 'prop-types'
import pick from 'ramda/src/pick'
import React from 'react'
import Panel from 'react-bootstrap/lib/Panel'
import Tab from 'react-bootstrap/lib/Tab'
import Tabs from 'react-bootstrap/lib/Tabs'
import Icon from 'react-fa'
import { connect } from 'react-redux'

import { setOptionsKey } from '../actions'
import CustomOptions from '../options/CustomOptions'
import DicewareOptions from '../options/DicewareOptions'
import GenericOptions from '../options/GenericOptions'

const options = {
  generic: {
    Component: GenericOptions,
    icon: 'check-square'
  },
  diceware: {
    Component: DicewareOptions,
    icon: 'book'
  },
  custom: {
    Component: CustomOptions,
    icon: 'question-circle'
  }
}

export function Options({ optionsKey, setOptionsKey }) {
  return (
    <Tabs activeKey={optionsKey} onSelect={setOptionsKey} id="options">
      {entries(options).map(([name, { Component, icon }]) => (
        <Tab key={name} eventKey={name} title={<div><Icon name={icon} /> {startCase(name)}</div>}>
          <Panel><Component /></Panel>
        </Tab>
      ))}
    </Tabs>
  )
}

Options.propTypes = {
  optionsKey: PropTypes.string.isRequired,
  setOptionsKey: PropTypes.func.isRequired
}

export const mapStateToProps = pick(['optionsKey'])

export default connect(mapStateToProps, { setOptionsKey })(Options)
