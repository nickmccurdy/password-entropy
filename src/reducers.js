import GenericOptions from './options/GenericOptions'
import { assoc, identity, map, T } from 'ramda'
import { combineReducers } from 'redux'

function createReducer(actionType, initialState, { callback = identity, validator } = {}) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = !validator || validator(payload)

    return matchesAction && isValid ? callback(payload, state) : state
  }
}

export default combineReducers({
  length: createReducer('SET_LENGTH', 6, { validator: n => n >= 1 }),
  options: combineReducers({
    custom: createReducer('SET_CUSTOM', 1, { validator: n => n >= 0 }),
    generic: createReducer ('TOGGLE_GENERIC', map(T, GenericOptions.toggles), {
      callback: ({ checked, name }, state) => assoc(name, checked, state)
    })
  }),
  optionsKey: createReducer('SET_OPTIONS_KEY', 'diceware')
})
