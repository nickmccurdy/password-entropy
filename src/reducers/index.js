import { combineReducers } from 'redux'

import custom from './custom'
import generic from './generic'
import length from './length'
import optionsKey from './optionsKey'

export default combineReducers({
  length,
  options: combineReducers({ custom, generic }),
  optionsKey
})
