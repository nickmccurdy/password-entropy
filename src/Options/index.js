import { Card, CardContent, TextField, Typography } from '@material-ui/core'
import PropTypes from 'prop-types'
import React from 'react'

import Custom from './Custom'
import Generic from './Generic'

export default function Options({
  length,
  onCustom,
  onGeneric,
  onLength,
  options,
  optionsKey
}) {
  return (
    <Card>
      <CardContent>
        <Typography variant="title" gutterBottom>
          Options
        </Typography>

        <TextField
          label="Number of Items"
          value={length}
          onChange={onLength}
          type="number"
          min="0"
          autoFocus
          fullWidth
        />

        {
          {
            Generic: (
              <Generic onGeneric={onGeneric} toggles={options.Generic} />
            ),
            Diceware: null,
            Custom: <Custom custom={options.Custom} onCustom={onCustom} />
          }[optionsKey]
        }
      </CardContent>
    </Card>
  )
}

Options.propTypes = {
  length: PropTypes.number.isRequired,
  onCustom: PropTypes.func.isRequired,
  onGeneric: PropTypes.func.isRequired,
  onLength: PropTypes.func.isRequired,
  options: PropTypes.shape({
    Custom: PropTypes.number.isRequired,
    Generic: PropTypes.shape({
      'Lowercase Letters': PropTypes.bool.isRequired,
      'Uppercase Letters': PropTypes.bool.isRequired,
      Numbers: PropTypes.bool.isRequired,
      Symbols: PropTypes.bool.isRequired
    }).isRequired
  }).isRequired,
  optionsKey: PropTypes.string.isRequired
}
