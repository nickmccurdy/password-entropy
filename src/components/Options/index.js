import book from '@fortawesome/fontawesome-free-solid/faBook'
import checkSquare from '@fortawesome/fontawesome-free-solid/faCheckSquare'
import questionCircle from '@fortawesome/fontawesome-free-solid/faQuestionCircle'
import Icon from '@fortawesome/react-fontawesome'
import classnames from 'classnames'
import startCase from 'lodash/startCase'
import entries from 'object.entries'
import PropTypes from 'prop-types'
import React, { Fragment } from 'react'
import { connect } from 'react-redux'
import { Card, CardBody, Nav, NavItem, NavLink, TabContent, TabPane } from 'reactstrap'

import Custom from './Custom'
import Diceware from './Diceware'
import Generic from './Generic'
import { selector, setOptionsKey } from '../../reducers/optionsKey'

const options = {
  generic: {
    component: <Generic />,
    icon: <Icon icon={checkSquare} />
  },
  diceware: {
    component: <Diceware />,
    icon: <Icon icon={book} />
  },
  custom: {
    component: <Custom />,
    icon: <Icon icon={questionCircle} />
  }
}

export function Options({ optionsKey, setOptionsKey }) {
  return (
    <Fragment>
      <Nav tabs>
        {entries(options).map(([name, { icon }]) => (
          <NavItem key={name}>
            <NavLink className={classnames({ active: optionsKey === name })} href="#" onClick={setOptionsKey.bind(null, name)}>
              {icon} {startCase(name)}
            </NavLink>
          </NavItem>
        ))}
      </Nav>

      <TabContent activeTab={optionsKey}>
        {entries(options).map(([name, { component }]) => (
          <TabPane key={name} tabId={name}>
            <Card><CardBody>{component}</CardBody></Card>
          </TabPane>
        ))}
      </TabContent>
    </Fragment>
  )
}

Options.propTypes = {
  optionsKey: PropTypes.string.isRequired,
  setOptionsKey: PropTypes.func.isRequired
}

export default connect(selector, { setOptionsKey })(Options)
