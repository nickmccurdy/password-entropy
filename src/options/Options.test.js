import { shallow } from 'enzyme'
import { capitalize, mapStateToProps, Options } from './Options'
import React from 'react'
import reducers from '../reducers'
import { createStore } from 'redux'

test('capitalize', () => {
  expect(capitalize('word')).toEqual('Word')
})

test('Options', () => {
  const setOptionsKey = jest.fn()
  const wrapper = shallow(<Options optionsKey="generic" setOptionsKey={setOptionsKey} />)
  wrapper.find('Uncontrolled(Tabs)').simulate('select')

  expect(wrapper).toMatchSnapshot()
  expect(setOptionsKey).toHaveBeenCalled()
})

test('mapStateToProps', () => {
  const state = createStore(reducers).getState()
  expect(mapStateToProps(state)).toEqual({ optionsKey: 'generic' })
})
