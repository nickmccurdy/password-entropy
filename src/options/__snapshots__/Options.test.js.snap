// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options 1`] = `
<Uncontrolled(Tabs)
  activeKey="diceware"
  id="options"
  onSelect={[Function]}
>
  <Tab
    eventKey="diceware"
    title={
      <div>
        <Icon
          Component="span"
          name="book"
        />
         
        Diceware
      </div>
    }
  >
    <Component />
  </Tab>
  <Tab
    eventKey="custom"
    title={
      <div>
        <Icon
          Component="span"
          name="question-circle"
        />
         
        Custom
      </div>
    }
  >
    <Connect(CustomOptions) />
  </Tab>
  <Tab
    eventKey="generic"
    title={
      <div>
        <Icon
          Component="span"
          name="check-square"
        />
         
        Generic
      </div>
    }
  >
    <Connect(GenericOptions) />
  </Tab>
</Uncontrolled(Tabs)>
`;
